docs/development/pigs/pig-016.rst

```diff
@@ -0,0 +1,215 @@
+  gammapy.cube.
+- Clean up gammapy.utils sub-packages
+
+After Gammapy v1.0, we propose to continue discussion on other changes
+considered while writing this PIG, described in the outlook section.
+
+Status
+======
+
+The current Gammapy package structure (as of v0.13):
+
+.. image:: pig-016-gammapy-package-organisation-status.png
+   :alt: Gammapy package structure status (v0.13)
+
+This illustrates the namespaces where the Gammapy API is exposed. The number of
+Python files is much larger (about 100) and not important for the end-user API.
+
+Proposal
+========
+
+We propose to introduce gammapy.modeling and gammapy.modeling.models, to
+dissolve gammapy.background and gammapy.image in the v0.14 development cycle,
+resulting in the following package structure for Gammapy v0.14 and v1.0:
+
+.. image:: pig-016-gammapy-package-organisation-proposal.png
+   :alt: Gammapy package structure proposal (v0.14)
+
+The following sections explain and motivate the proposed changes.
+
+Introduce gammapy.modeling
+--------------------------
+
+Currently our fitting and modeling framework is in gammapy.utils.fitting,
+and the built-in models in Gammapy are distributed like this:
+
+- gammapy.spectrum.models -- 1D spectral models
+- gammapy.image.models -- 2D spatial models
+- gammapy.cube.models -- 3D cube models
+- gammapy.time.models -- 1D time models
+
+Our proposal is to create a new sub-package gammapy.modeling and
+gammapy.modeling.models following the organisation chosen for models in Astropy and
+restructure code like this:
+
+- gammapy/utils/fitting -> gammapy/modeling/fitting
+- gammapy/utils/serialisation -> gammapy/modeling/serialisation
+- gammapy/spectrum/models -- gammapy/modeling/models/spectrum
+- gammapy/image/models -- gammapy/modeling/models/image
+- gammapy/cube/models -- gammapy/modeling/models/cube
+- gammapy/time/models -- gammapy/modeling/models/time
+
+Since it is very inconvenient for users to have to type sub-sub-sub-package
+names, we propose to expose the end-user API direction in gammapy.modeling for
+general classes (Parameter, Parameters, Model, Fit, Dataset). For the models, we
+propose to expose them in gammapy.modeling.models (same as Astropy), and to
+introduce a factory function to make it easy to make models that is easy to
+import and use. E.g. it could be gammapy.modeling.make_model and take a model
+specification string or dictionary.
+
+We propose to add "Source" to all (currently 3D) source models, and to add
+"Spectral", "Spatial" and "Time" for the spectral, spatial and time models. This
+avoids user confusion concerning the type of model, which currently exists in
+v0.13, because for some models (Gaussian, diffuse, table, constant) it isn't
+obvious what kind they are.
+
+The motivation for this change is weak, but in discussions at the Erlangen
+coding sprint most Gammapy developers preferred it. The reasoning was that it's
+a bit easier to find and maintain models if they are all in one place.
```

Comment:
I would mention here that the drawback is very long names, e.g. `SpectralPowerLaw` is probably still fine but `SpectralExponentialCutoffPowerLaw3FGL` is a real mess.

Reply:
I've added the sentence "The drawback is that this generates very long class
names, such as e.g. SpectralExponentialCutoffPowerLaw3FGL."